rfcs/20200624-pluggable-device-for-tensorflow.md

```diff
@@ -0,0 +1,251 @@
+
+1. `PluggableDeviceFactory` creates and initializes a set of `PluggableDevice` instances when the session is created.  
+```cpp
+   PluggableDeviceFactory::CreateDevices(SessionOptions& options, const string& name_prefix, std::vector<std::unique_ptr<Device>>* devices) {
+     for (int i = 0; i < options.device_count(); i++) {
+      PluggableDevice pluggable_device = CreatePluggableDevice(options,i); //set allocator
+      pluggable_device->Init(options);
+      devices.push_back(std::move(pluggable_device));
+     }
+   }
+```
+
+2. `PluggableDevice` object binds a StreamExecutor and creates a set of Streams during the initialization.Streams include one compute stream and several memory copy streams.
+```cpp
+   void PluggableDevice::Init(SessionOption& options) {  
+     se::Platform* platform= se::MultiPlatformManager::PlatformWithName("PluggableDevice");
+     stream_executor_ = platform->ExecutorForDevice(pluggable_dev_id_);
+     compute_stream_ = new se::Stream(stream_executor_);
+     compute_stream_->Init();
+     host_to_device_stream_ = new se::Stream(stream_executor_);
+     host_to_device_stream_->Init();
+     ...
+   }  // create StreamExecutor
+```
+3. `PluggableDevicePlatform` is responsible for the StreamExecutor creation. It creates an `SE_StreamExecutor` and `SE_Device` object through create_stream_executor and create_device which are registered in the `SE_Platform`. Then `PluggableDeviceExecutor` is constructed with `SE_StreamExecutor` and `SE_Device` object.   
+```cpp
+   StreamExecutor* PluggableDevicePlaform::ExeutorForDevice(int device_id） {
+    auto config = get_plugin_config(device_id);
+    SE_Options* se_option = get_se_option(device_id);
+    SE_StreamExecutor* se= platform_->create_stream_executor();
+    SE_Device* sd = platform_->create_device(se_options)
+    auto executor = absl::make_unique<StreamExecutor>(this, absl::make_unique<PluggableDeviceExecutor>(config, se, sd));
+    return std::move(executor);
+   }
+```
+**TensorFlow Proper**
+
+TensorFlow proper needs to be extended to support a new class `PluggableDevice` to represent a set of new third-party devices and a new stream executor platform (`PluggableDevicePlatform`) to create the device and related resources with the information registered from plugin. 
+
+Two sets of classes need to be defined in TensorFlow proper. 
+* Set 1: `PluggableDevice` related classes 
+   * class `PluggableDevice`:  a class represents a set of new third-party devices, it has a new device type named "PluggableDevice"/DEVICE_PLUGGABLE.
+   * class `PluggableDeviceFactory`: a device factory to create the PluggableDevice
+   * class `PluggableDeviceBFCAllocator`: a PluggableDevice memory allocator that implements a ‘best fit with coalescing’ algorithm.
```

Comment:
Is it going to extend BFCAllocator: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/bfc_allocator.h#L49 ?

Reply:
yes, we want to extend it similar as GPUBFCAllocator for legacy GPU option support: https://github.com/tensorflow/tensorflow/blob/cbb94efa589fc1d1b34948b9ca03829fe8613d12/tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc#L22